### PR TITLE
[WIP] Smooth(er) movement

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request
https://www.youtube.com/watch?v=T57yYpU3OHs&feature=youtu.be
Ports CEV Eris's glide_size algorithm to smooth out movement.
Fixes the janky movement at full speed at 60fps.
Note that this doesn't affect the actual movespeed or move mechanics, the same delays apply, this just changes how it looks.

## Why It's Good For The Game

Looks nice.
Easier to tell if someone is affected by movement slowdown from a single movement.

## Changelog
:cl:
tweak: Changed movement animation slightly to help it look a bit smoother. 60 fps now works properly.
/:cl:

The math is a little off, I'm still working on solving that. It isn't as smooth as it could be. Probably going to need to write my own algorithm.
I also need to see if I can make diagonals a little better, the different move delay on them fucks with it a tiny bit.

### CODE REVIEW

**Comment by optimumtact** [comment]
can't wait for the horde of arguments about it being too floaty or fine

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -196,6 +196,15 @@
 	if(pulledby && moving_diagonally != FIRST_DIAG_STEP && get_dist(src, pulledby) > 1)		//separated from our puller and not in the middle of a diagonal move.
 		pulledby.stop_pulling()
 
+/atom/movable/proc/set_glide_size(target = 8)
+	glide_size = target
+
+	for(var/atom/movable/AM in contents)
```
Comment (optimumtact) [inline code/game/atoms_movable.dm]: why are you recursing into contents.

most items that are in contents aren't visually viewable on the turf

Comment (actioninja) [reply code/game/atoms_movable.dm]: good point, this was how it was done in eris but they might use contents for buckled as well.

**Comment by WJohn** [comment]
Oooh! This makes walking or slowness MUCH more obvious! I like this for that alone.

Does this fix dragging the sloth, by the way?

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -196,6 +196,12 @@
 	if(pulledby && moving_diagonally != FIRST_DIAG_STEP && get_dist(src, pulledby) > 1)		//separated from our puller and not in the middle of a diagonal move.
 		pulledby.stop_pulling()
 
+/atom/movable/proc/set_glide_size(target = 8)
+	glide_size = target
+
+	for(var/atom/movable/AM in buckled_mobs)
+		AM.set_glide_size(target)
```
Comment (AnturK) [inline code/game/atoms_movable.dm]: So since buckling doesn't update speed, the buckled mobs will be broken until they change speed ?

**Comment by MrPerson** [comment]
Diagonal movements are sqrt(world.icon_size ^ 2 + world.icon_size ^ 2) for the numerator instead of just world.icon_size, thanks Pythagoras.

**Comment by Shapsy** [comment]
I believe MSO tried something similar when he was working on the MC, I'm not a fan of it since while it looks nice and smooth, it makes it harder to stop or make turns on specific tiles, especially at low speeds. At least with snap movement, you can be more sure of what tile the game considers you on

File: code/game/atoms_movable.dm (hunk 2 above)
Comment (actioninja) [reply code/game/atoms_movable.dm]: yep, not anymore though.

**Comment by actioninja** [comment]
> Oooh! This makes walking or slowness MUCH more obvious! I like this for that alone.
> 
> Does this fix dragging the sloth, by the way?

It does now.

**Comment by PKPenguin321** [comment]
So in layman's terms, this makes it so instead of slowdown being "go -> cooldown stop -> go -> cooldown  stop," you instead actually move slower between tiles? That's pretty cool
What happens if you get stunned while moving between tiles or something?

**Comment by actioninja** [comment]
@PKPenguin321 yes
As stated, this doesn't actually change the mechanics of how moving works. As soon as you start the move, you're already considered to be on the next tile as far as the game cares. Your sprite is still used for stuff like click checks and hitboxes but all other stuff like movement checks, atmos checks, etc are tile based and you're already "on" the tile you're moving to.

**Comment by PKPenguin321** [comment]
Your sprite still being your hitbox actually has pretty severe balance implications... But tbh I like the visual effect enough to overlook that

**Comment by actioninja** [comment]
Eh, slow moving vs completely locked in place for half a second. It's a trade off. I don't think it would have *huge* balance implications even if it does change things.

**Comment by MrPerson** [comment]
This removes tactical stutterstepping. Yes I'm serious.

For real though the whole "can't stop on a dime as easily" may turn out to be a real issue. Of course if so, there's some other ideas to try out rather than just declaring this a failure.

**Comment by actioninja** [comment]
> This removes tactical stutterstepping. Yes I'm serious.

I know what you mean and I've seen some people get super tricky with this. I actually wasn't considering that while making this. I'd consider it an acceptable loss, but I'll also be honest and say I'm not a huge fan of it to begin with because it's a very "gamey" mechanic.

**Comment by Bawhoppen** [comment]
We can't stop progress for a few measly balance concerns :+1:

**Comment by actioninja** [comment]
wait I forgot to fix pulling

**Comment by actioninja** [comment]
I also need to figure out how to handle pushing as well.

**Comment by actioninja** [comment]
On some further investigation, getting pushing working is going to be a headache I don't want to deal with right now. I'll revisit this later.
